GitCommands/Git/GitModule.cs

```diff
@@ -2136,50 +2136,60 @@ IReadOnlyList<Remote> ParseRemotes(IEnumerable<string> lines)
             {
                 var remotes = new List<Remote>();
 
-                /*
-                 * $ git remote -v
-                 * origin  git@github.com:drewnoakes/gitextensions.git (fetch)
-                 * origin  git@github.com:drewnoakes/gitextensions.git (push)
-                 * upstream        git@github.com:gitextensions/gitextensions.git (fetch)
-                 * upstream        git@github.com:gitextensions/gitextensions.git (push)
-                 *
-                 * That's a tab character between the first two records. There's a space before (fetch/push).
-                 */
+                // See tests for explanation of the format
 
                 using (var enumerator = lines.GetEnumerator())
                 {
                     while (enumerator.MoveNext())
                     {
                         var fetchLine = enumerator.Current;
-
-                        // An invalid module is not an error; we simply return an empty list of remotes
                         if (fetchLine.IndexOf("not a git repository", StringComparison.OrdinalIgnoreCase) >= 0)
                         {
+                            // An invalid module is not an error; we simply return an empty list of remotes
                             return remotes;
                         }
 
+                        var fetchMatch = _remoteVerboseLineRegex.Match(fetchLine);
+                        if (!fetchMatch.Success
+                            || (fetchMatch.Groups["direction"].Value != "fetch"
+                               && fetchMatch.Groups["direction"].Value != "push"))
+                        {
+                            // Ignore malformed and unknown entries
+                            continue;
+                        }
+
+                        var name = fetchMatch.Groups["name"].Value;
+                        var fetchUrl = fetchMatch.Groups["url"].Value;
+                        if (fetchMatch.Groups["direction"].Value == "push")
+                        {
+                            if (remotes.Count <= 0 || name != remotes[remotes.Count - 1].Name)
+                            {
+                                throw new Exception("Unable to update remote pushurl for command output: " + fetchLine);
+                            }
+
+                            var updatedPushUrl = fetchUrl;
```

Comment:
I don't understand what we do here.

Do we replace the previously found push url because we support only one?

If, yes, I tend to think that it is the first one the more important.
So we could keep it and fill a collection with additional push url. That we don't handle in the rest of the application for the moment but will let us display text like "+ 1 other remote" somewhere (push form?)

That would looks like `remotes[remotes.Count - 1].AddOtherPushUrl(updatedPushUrl)`...

Please add at least a comment.

Reply:
I did not bother with making PushUrl a List (see description)
The last url is what you get when requesting with git-config (for instance in FormRemote), this aligns with what is displayed in FormRemote
Will add a comment, unless other changes are requested

Reply:
Are we talking about this?
![image](https://user-images.githubusercontent.com/4403806/66215810-a7f2b800-e707-11e9-808a-d5816dfcf0a8.png)

If so, it can be quite useful.

I think we need to make it a list and read everything a user may have configured.
We'll need to update the UI to match as well (can be a separate PR)

Reply:
> Are we talking about this?

Yes, but this remote parsing is not used (but config parsing)
Adding multiple push url in the GUI is definitely an enhancement in a separate PR (nothing I am interested in), but I could make PushUrl a list if requested here.

Reply:
That's a code that is not easy to understand, so I think that we will not change it often.

That's why I think we could do it now as we are in it. And I even think that it will made the code clearer.

But that is just my opinion and maybe a comment is enough. I don't what to force someone to do something...

Reply:
> But that is just my opinion and maybe a comment is enough. I don't what to force someone to do something...

When implementing, you have to draw the line somewhere, everything cannot be polished so it shines. In this case I really stopped when detecting a second pushurl, but then I found that the config extraction used the last so I changed the implementation. 
If you as a reviewer was not happy about it, of course you can request changes, this is reasonable. 

Also changed order so fetchurl is before pushurl(s)
I will squash when approved

Reply:
Let's enumerate all configured remotes (fetch and as many pushes) as user may have configured.
The UI is still limited but we can deal with this later.

Reply:
Brief check of usage of pushurl:
Most usage of remotes is using GitCommands.Remotes.ConfigFileRemote that reads/write config SettingKeyString.RemotePushUrl, a single pushurl (last in the list)

If the multiple pushurl is a desired feature, Remote cannot replace ConfigFileRemote  directly as ConfigFileRemote  handles disabled remotes and PuttyKeys too. So disabling a remote will not display the extra push urls (but the urls will be displayed when enabling again).

Usages of ConfigFileRemote :
ExternalLinkRevisionParser.cs
GitUI\BranchTreePanel\RepoObjectsTree.Nodes.Remotes.cs
FormPull.cs
FormPush.cs (OK)
FormRemotes.cs
CommitInfo.cs (for ExternalLinkRevisionParser)


 Most is filled in by reading from config

Explicit use of pushurl from config also in FormClone: Should be OK

GitUI\CommandsDialogs\SettingsDialog\Pages\BuildServerIntegrationSettingsPage.cs
prioritizes a (single) push url over the fetchurl, which should be incorrect

Some usages like GitHub3Plugin.cs lists remote names, then the remote url

So a second update is required before multiple push url can be displayed in FormRemote

I have pushed some comments

Reply:
Thank you for the analysis.

I backtrack on my requests, because it is clearly not going to be beneficial - other parts of the app won't be able to support or utilise a list of push remotes.
So I leave it up to you to make the best decision here.

Given that you've done the analysis and (hopefully) understand the breadth of the problem, do you think you could raise an issue (or more) describing what you think needs to be done to support multiple push remotes and provide necessary UI/UX for this?

Reply:
> could raise an issue

That is the plan, I just put down the analysis here so I do not forget about it.
I do not find the feature to be important in an GUI, but the inconsistencies in the data structures is itching.

Unless someone has better ideas, I will review the comments in ConfigFileRemote  when adding the issue, before merging.

Reply:
See related issue and PR.
RepoObjectsTree.Nodes.Remotes.cs uses both Remote.cs and ConfigFileRemote

Reply:
@pmiossec 
Can you please review again?